src/app/components/breadcrumbs/service/terra-breadcrumbs.service.ts

```diff
@@ -0,0 +1,424 @@
+        let container:TerraBreadcrumbContainer = this._containers[urlPartsCount];
+        if(isNullOrUndefined(container))
+        {
+            container = new TerraBreadcrumbContainer();
+            this._containers.push(container);
+        }
+
+        // search for existing breadcrumb
+        let breadcrumb:TerraBreadcrumb = container.breadcrumbList.find((bc:TerraBreadcrumb) =>
+        {
+            return bc.routerLink === url;
+        });
+
+        // breadcrumb not found
+        if(isNullOrUndefined(breadcrumb))
+        {
+            let label:string = this.getBreadcrumbLabel(route);
+            let currentContainerIndex:number = this._containers.indexOf(container);
+            let previousContainer:TerraBreadcrumbContainer = this._containers[currentContainerIndex - 1];
+            let parentBreadcrumb:TerraBreadcrumb = isNullOrUndefined(previousContainer) ? undefined : previousContainer.currentSelectedBreadcrumb;
+            breadcrumb = new TerraBreadcrumb(label, parentBreadcrumb, url);
+            container.breadcrumbList.push(breadcrumb);
+        }
+
+        // select breadcrumb and update visibilities
+        container.currentSelectedBreadcrumb = breadcrumb;
+        this.updateBreadcrumbVisibilities(container, breadcrumb.parent);
+    }
+
+    private getBreadcrumbLabel(route:Route):string
+    {
+        let label:string = '';
+        if(!isNullOrUndefined(route.data))
+        {
+            if(typeof route.data.label === 'function')
+            {
+                let activatedSnapshot:ActivatedRouteSnapshot = this.findActivatedRouteSnapshot(this.router.routerState.snapshot.root);
+
+                label = route.data.label(this.translation, activatedSnapshot.params, activatedSnapshot.data);
+            }
+            else
+            {
+                label = this.translation.translate(route.data.label);
+            }
+        }
+        return label;
+    }
+
+    private updateBreadcrumbVisibilities(breadcrumbContainer:TerraBreadcrumbContainer, parentBreadcrumb:TerraBreadcrumb):void
+    {
+        breadcrumbContainer.breadcrumbList.forEach((bc:TerraBreadcrumb) =>
+        {
+            bc.isHidden = bc.parent !== parentBreadcrumb || !isNullOrUndefined(bc.parent) && bc.parent.isHidden;
+        });
+
+        // is the current selected breadcrumb now hidden?
+        if(breadcrumbContainer.currentSelectedBreadcrumb.isHidden)
+        {
+            // search for another breadcrumb to be selected that is not hidden
+            let foundBreadcrumb:TerraBreadcrumb = breadcrumbContainer.breadcrumbList.find((bc:TerraBreadcrumb) => !bc.isHidden);
+            if(!isNullOrUndefined(foundBreadcrumb))
+            {
+                breadcrumbContainer.currentSelectedBreadcrumb = foundBreadcrumb;
+            }
+        }
+    }
+
+    // same exists in TerraRouterHelper
+    private findActivatedRouteSnapshot(activatedRouteSnapshot:ActivatedRouteSnapshot):ActivatedRouteSnapshot
```

Comment:
in einen Helper (ActivatedRouteHelper oder RouteHelper) auslagern.. So können wir Stück für Stück die Funktionalitäten aus dem TerraRouterHelper hierher überführen.